cpp/ql/src/Likely Bugs/Arithmetic/SignedOverflowCheck.ql

```diff
@@ -0,0 +1,26 @@
+/**
+ * @name Undefined result of signed test for overflow
+ * @description Testing for overflow by adding a value to a variable
+ *              to see if it "wraps around" works only for
+ *              unsigned integer values.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id cpp/signed-overflow-check
+ * @tags reliability
+ *       security
+ */
+
+import cpp
+private import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+private import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+
+from RelationalOperation ro, AddExpr add, Expr expr1, Expr expr2
+where
+  ro.getAnOperand() = add and
+  add.getAnOperand() = expr1 and
+  ro.getAnOperand() = expr2 and
+  globalValueNumber(expr1) = globalValueNumber(expr2) and
+  add.getUnspecifiedType().(IntegralType).isSigned() and
+  exprMightOverflowPositively(add)
```

Comment:
I'll re-run the query in the LGTM query console, but I haven't seen any cases with a negative `delta`.

Reply:
https://lgtm.com/query/4231059764463143298/

Note that if `delta` is negative and sufficiently small, we may have to contend with negative overflow.  Perhaps we can use `exprMightOverflowNegatively()` as well?

Reply:
Thanks for sharing the query console link. I don't see any results that would benefit from a check that `delta` is non-negative. Do you?

I do see some dubious results that the query will need to exclude. An easy starting point would be to filter out results that come from macro expansions.

I agree that in principle we should add `exprMightOverflowNegatively`. Do you then want to tie the overflow direction to the direction of the comparison operator?

Reply:
> Thanks for sharing the query console link. I don't see any results that would benefit from a check that `delta` is non-negative. Do you?

No, I do not either.

> I do see some dubious results that the query will need to exclude. An easy starting point would be to filter out results that come from macro expansions.

I agree.  I've analyzed the results a bit more closely, and seems that the query always flags the problem correctly.  But the layer of obfuscation created by a macro renders our query rather useless.

> I agree that in principle we should add `exprMightOverflowNegatively`. Do you then want to tie the overflow direction to the direction of the comparison operator?

I don't think we should involve the direction of operators, since they're always two sides of the same coin, i.e., one can use `<` to test for overflow or `>=` to test for _absence_ of overflow.

But another question comes to mind: should we handle subtraction/underflow?  The use of  `exprMightOverflowNegatively` would make sense in that context.

Reply:
Re-ran the 73 project suite with `exprMightOverflowPositively` `or` `exprMightOverflowNegatively` and did not unearth anything further.